### PR TITLE
SWDEV-102314 "[OCL-LC-ROCm] AMD lower kernel calls"

### PR DESCRIPTION
**Pull request by dfukalov**
Ported lower kernels calls pre-link pass from hsail compiler

### CODE REVIEW

File: lib/Target/AMDGPU/AMDGPULowerKernelCalls.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  return NewF;
+}
+
+static Function *cloneKernel(Function &F) {
+  ValueToValueMapTy ignored;
+  Function *NewF =
+      F.empty() ? copyDeclaration(F) : CloneFunction(&F, ignored, false);
+  F.getParent()->getOrInsertFunction(NewF->getName(), NewF->getFunctionType(),
+    NewF->getAttributes());
+  NewF->setCallingConv(CallingConv::C);
```
Comment (vpykhtin) [inline lib/Target/AMDGPU/AMDGPULowerKernelCalls.cpp:74]: Why not just getOrInsertFunction with the new name from setNameForBody?

**Comment by vpykhtin** [comment]
LGTM.

**Comment by tstellarAMD** [comment]
Are you sure this is necessary?  I thought the compiler already did this.

**Comment by tstellarAMD** [comment]
There is code in AMDGPUAlwaysInlinePass, which is meant to do this.  If that code isn't working, I would prefer to fix that rather than add a new pass.

**Comment by dfukalov** [comment]
There is and ability in OpenCL to call kernel function from other kernel or function. The pass finds such calls, create clones for callees (with mangled name and removed amdgpu_kernel calling convention) and patches calls to such functions.
As I understand, it's a bit different from cloning part of AMDGPUAlwaysInline pass. Are you sure processing of kernel functions calls functionality should be introduced to the inliner?

**Comment by tstellarAMD** [comment]
Ok, I forgot we would need this for when function calls were supported.  Did you test this with the OCL conformance tests?

**Pull request merged**